### DISCUSSION TITLE
How do I use template variables?

### DISCUSSION THREAD

**Discussion by ddahan** [meteorlxy/vue-showdown#33] [Q&A]
How do you allow VueShowndown to interpret template variables on the fly?

For example, with this single file component:

```js
<template>
      <div>
        <VueShowdown class="prose" :markdown="content" :vueTemplate="true" />
      </div>
</template>

<script>
import { Urls } from '../consts/urls.consts';

export default {
  data() {
    return {
      content: 'Write some `Markdown` here :kissing:',
      myDynamicUrl: Urls.apiHost,
    };
  },
};
</script>
```

How can I process so that text like `{{myDynamicUrl}}` is rendered by VueShowdown?

![image](https://user-images.githubusercontent.com/1933516/116048937-05e24a00-a676-11eb-87a1-a9161ad7059a.png)

**Comment by meteorlxy** [comment]
The `vueTemplate` support is quite minimal now, so I have to say you can't use it as that.

But it's possible to support it. What about usage like this:

```vue
<template>
  <div>
    <VueShowdown class="prose" :markdown="content" :vue-template="true" :vue-template-data="{ myDynamicUrl }" />
  </div>
</template>

<script>
import { Urls } from '../consts/urls.consts';

export default {
  data() {
    return {
      content: 'Write some `Markdown` here :kissing:',
      myDynamicUrl: Urls.apiHost,
    };
  },
};
</script>
```

**Reply by ddahan** [reply]
That would be great! The trick here seems to be that the text `{{myDynamicUrl}}` will need to be interpreted directed from the markdown content.

```js
export default {
  data() {
    return {
      content: 'Write some `Markdown` here :kissing: and hope that {{myDynamicUrl}} is interpreted at runtime',
      myDynamicUrl: Urls.apiHost,
    };
  },
};
```

If you know how do it, it would be amazing! :)

**Reply by ddahan** [reply]
FYI, to understand the need, I'm building a blog engine, and articles are stored in Markdown in a database.
The problem is that for media (images for example), I need dynamic urls to match environment host address (like `http://127.0.0.1` in local, and `https://myrealsite.com` in production).

Otherwise I'm forced to embbed base64 images, that creates many other issues.
That's why having the ability to inject variables in my Markdown content, which will be interpreted at runtime, would be neat!

**Reply by meteorlxy** [reply]
I've tested locally and things work well.

BTW, are you using vue2 or vue3 ? 🤔  (I'm a little lazy and do not want to add this feature to v2 version if you don't need it lol)

**Reply by ddahan** [reply]
Vue3, please keep being lazy :) :)

**Reply by meteorlxy** [reply]
Supported in v3.1.0 🚀

**Reply by ddahan** [reply]
That was ... fast!!! Thanks! I'll try it as soon as possible!

**Reply by ddahan** [reply]
I succeed in making it work in body, but not inside MD attribute. Any idea?

![image](https://user-images.githubusercontent.com/1933516/116099616-6db38780-a6ac-11eb-848e-225130267fc8.png)

**Reply by meteorlxy** [reply]
Will, it should be valid markdown first, then vue template.

The `![]({{ foo }})` is not a valid markdown syntax, that's the reason why it did not work.

Try using like this (have updated the wrong code): 

```md
<img :src="publicBlogUrl + '/discounts.jpeg'">
```

Anyway, try using vue syntax directly instead of combining with markdown syntax

**Reply by meteorlxy** [reply]
An alternative way is to create a showdown plugin to handle such syntax. It's out of the scope of this project, though

**Reply by ddahan** [reply]
I have a similar issue:

## before using img
<img width="1303" alt="Screenshot 2021-04-26 at 16 56 46" src="https://user-images.githubusercontent.com/1933516/116104258-96d61700-a6b0-11eb-9d92-5805d379bb39.png">

## after using img
<img width="1304" alt="Screenshot 2021-04-26 at 16 57 38" src="https://user-images.githubusercontent.com/1933516/116104333-a5243300-a6b0-11eb-9c3d-f2ecad502631.png">


(simplified) component:

```js
<template>
    <VueShowdown
      class="prose"
      :markdown="content"
      :vueTemplate="true"
      :vueTemplateData="{ blogUrl }"
    />
</template>

<script>
import { Urls } from '../consts/urls.consts';

export default {
  data() {
    return {
      blogUrl: Urls.blog,
      content: 'Write some `Markdown` here :kissing:',
    };
  },
};
</script>
```

**Reply by meteorlxy** [reply]
You'd better check the console for what's the error, or check the DOM structure for what is rendered

**Reply by meteorlxy** [reply]
Ah damn. What I wrote is not valid markdown syntax too.

Try this instead:

```md
<img :src="publicBlogUrl + '/discounts.jpeg'">
```

**Reply by meteorlxy** [reply]
Anyway, although vue template syntax is somehow compatible with markdown, it still have some limitations. We have similar mechanism in vuepress, you may have interests about it (if you don't know the vuepress project yet)

**Reply by ddahan** [reply]
Indeed I noticed some interaction issues because of some Django code that uses the same `{{`
I deactivated the option when it's not required. Thanks again for your help and the quality of your tool ;)
